File: Algorithms/C++/224-Basic_Calculator.cpp
// Runtime: 12 ms, faster than 88.32% of C++ online submissions for Basic Calculator.
// Memory Usage: 10.3 MB, less than 100.00% of C++ online submissions for Basic Calculator.
class Solution {
public:
    int calculate(string s) {
        int res = 0;
        stack<int> stac;
        long int num = 0;
        int sign = 1;
        for (char c : s) {
            if ('0' <= c && c <= '9')
                num = num * 10 + c - '0';
            else if (c == '+') {
                res += sign * num;
                sign = 1;
                num = 0;
            } else if (c == '-') {
                res += sign * num;
                sign = -1;
                num = 0;
            } else if (c == '(') {
                stac.push(res);
                stac.push(sign);
                sign = 1;
                res = 0;
            } else if (c == ')') {
                res += sign * num;
                res *= stac.top();
                stac.pop();
                res += stac.top();
                stac.pop();
                num = 0;
            }
        }
        return res + sign * num;
    }
};

File: Algorithms/Python3.x/224-Basic_Calculator.py
# Runtime: 80 ms, faster than 95.68% of Python3 online submissions for Basic Calculator.
# Memory Usage: 15.3 MB, less than 7.14% of Python3 online submissions for Basic Calculator.
class Solution:
    def calculate(self, s: str) -> int:
        result = 0
        stack = []
        num = 0
        sign = 1  # 1 is +, -1 is -
        for c in s:
            if c.isdigit():
                num = num * 10 + int(c)
            elif c == '+':
                result += sign * num
                sign = 1
                num = 0
            elif c == '-':
                result += sign * num
                sign = -1
                num = 0
            elif c == '(':
                stack.append(result)
                stack.append(sign)
                sign = 1
                result = 0
            elif c == ')':
                result += sign * num
                result *= stack.pop()
                result += stack.pop()
                num = 0
            else:
                continue
        return result + sign * num
